src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs

```diff
@@ -36,202 +79,565 @@ public SafeX509ChainHandle SafeHandle
 
-                        // Ownership was transferred to the cert stack.
-                        handle.SetHandleAsInvalid();
+                    lastCert = currentHandle;
+
+                    ArraySegment<byte> authorityInformationAccess =
+                        OpenSslX509CertificateReader.FindFirstExtension(
+                            currentCert,
+                            Oids.AuthorityInformationAccess);
+
+                    if (authorityInformationAccess.Count == 0)
+                    {
+                        break;
                     }
 
-                    if (lookupCrl)
+                    X509Certificate2 downloaded = DownloadCertificate(
+                        authorityInformationAccess,
+                        ref _remainingDownloadTime);
+
+                    // The AIA record is contained in a public structure, so no need to clear it.
+                    ArrayPool<byte>.Shared.Return(authorityInformationAccess.Array);
+
+                    if (downloaded == null)
                     {
-                        CrlCache.AddCrlForCertificate(
-                            cert,
-                            store,
-                            revocationMode,
-                            verificationTime,
-                            ref remainingDownloadTime);
+                        break;
+                    }
 
-                        // If we only wanted the end-entity certificate CRL then don't look up
-                        // any more of them.
-                        lookupCrl = revocationFlag != X509RevocationFlag.EndCertificateOnly;
+                    if (downloadedCerts == null)
+                    {
+                        downloadedCerts = new List<X509Certificate2>();
                     }
+
+                    AddToStackAndUpRef(downloaded.Handle, _untrustedLookup);
+                    downloadedCerts.Add(downloaded);
+                    
+                    Interop.Crypto.X509StoreCtxRebuildChain(storeCtx);
+                    statusCode = Interop.Crypto.X509StoreCtxGetError(storeCtx);
                 }
+            }
 
-                if (revocationMode != X509RevocationMode.NoCheck)
+            if (statusCode == Interop.Crypto.X509VerifyStatusCode.X509_V_OK && downloadedCerts != null)
+            {
+                using (SafeX509StackHandle chainStack = Interop.Crypto.X509StoreCtxGetChain(_storeCtx))
                 {
-                    if (!Interop.Crypto.X509StoreSetRevocationFlag(store, revocationFlag))
+                    int chainSize = Interop.Crypto.GetX509StackFieldCount(chainStack);
+                    Span<IntPtr> tempChain = stackalloc IntPtr[DefaultChainCapacity];
+                    IntPtr[] tempChainRent = null;
+
+                    if (chainSize < tempChain.Length)
                     {
-                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                        tempChain = tempChain.Slice(chainSize);
```

Comment:
Is this supposed to be Slice(0, chainSize)?

Reply:
Yep, guess I've always been lucky and come out under half.